File: solutions/julia/protein-translation/1/protein-translation.jl

codon_dict = Dict("UGU" =>  "Cysteine",
        "UGC" =>  "Cysteine",
        "UUA" =>  "Leucine",
        "UUG" =>  "Leucine",
        "AUG" =>  "Methionine",
        "UUU" =>  "Phenylalanine",
        "UUC" =>  "Phenylalanine",
        "UCU" =>  "Serine",
        "UCC" =>  "Serine",
        "UCA" =>  "Serine",
        "UCG" =>  "Serine",
        "UGG" =>  "Tryptophan",
        "UAU" =>  "Tyrosine",
        "UAC" =>  "Tyrosine",
        "UAA" =>  "STOP",
        "UGA" =>  "STOP",
        "UAG" =>  "STOP")

 
function proteins(strand)    
    result = String[]
    if 0==length(strand)
        return result
    end
    codons = collect(Iterators.partition(strand, 3))
    for codon in codons
        if haskey(codon_dict, codon) 
            p = codon_dict[codon]
            if "STOP" == p
                break
            else
                push!(result,p)
            end
        else
            throw(DomainError(strand))
        end
    end
    result
end

File: solutions/clojure/protein-translation/1/src/protein_translation.clj
(ns protein-translation)

(def codon-map 
  {
                "AUG" "Methionine"
                "UUU" "Phenylalanine"
                "UUC" "Phenylalanine"
                "UUA" "Leucine"
                "UUG" "Leucine"
                "UCU" "Serine"
                "UCC" "Serine"
                "UCA" "Serine"
                "UCG" "Serine"
                "UAU" "Tyrosine"
                "UAC" "Tyrosine"
                "UGU" "Cysteine"
                "UGC" "Cysteine"
                "UGG" "Tryptophan"
                "UAA" "STOP"
                "UAG" "STOP"
                "UGA" "STOP"
   }
  )

(defn translate-codon [codon] ;; <- arglist goes here
      ;; your code goes here
  (if (not (= (count codon) 3)) (throw (IllegalArgumentException. "Must have three items in a Codon."))
    (get codon-map codon))
)

(defn translate-rna [rna-sequence] ;; <- arglist goes here
  ;; your code goes here
  (take-while #(not (= "STOP" %))
              (map translate-codon 
                   (map #(apply str %)(partition 3 rna-sequence)
                        )))
)
